File: clients/java-client/src/main/java/io/ehdev/conrad/client/java/internal/DefaultVersionManagerClient.java
package io.ehdev.conrad.client.java.internal;

import io.ehdev.conrad.client.java.Version;
import io.ehdev.conrad.client.java.VersionManagerClient;
import io.ehdev.conrad.client.java.VersionServiceConfiguration;
import io.ehdev.conrad.client.java.exception.UnsuccessfulRequestException;
import io.ehdev.conrad.model.repo.RepoCreateModel;
import io.ehdev.conrad.model.repo.RepoResponseModel;
import io.ehdev.conrad.model.version.UncommitedVersionModel;
import io.ehdev.conrad.model.version.VersionCommitModel;
import io.ehdev.conrad.model.version.VersionCreateModel;
import io.ehdev.conrad.model.version.VersionSearchModel;
import org.apache.http.client.HttpClient;
import org.eclipse.jgit.api.errors.GitAPIException;

import java.io.File;
import java.io.IOException;
import java.net.URISyntaxException;
import java.util.List;

public class DefaultVersionManagerClient implements VersionManagerClient {

    final HttpHelper httpHelper;
    final VersionServiceConfiguration configuration;

    public DefaultVersionManagerClient(HttpClient client, VersionServiceConfiguration configuration) {
        this.httpHelper = new HttpHelper(client, configuration);
        this.configuration = configuration;
    }

    public Version findFinalVersion(File searchDir) throws IOException, GitAPIException, URISyntaxException, UnsuccessfulRequestException {
        GitHelper gitHelper = new GitHelper(searchDir);
        return new Version(httpHelper.findFinalVersion(gitHelper.getHeadCommitId().getName()).getVersion());
    }

    @Override
    public Version findVersion(File searchDir) throws IOException, GitAPIException, URISyntaxException, UnsuccessfulRequestException {
        GitHelper gitHelper = new GitHelper(searchDir);
        List<String> commitIds = gitHelper.findCommitsFromHead();
        return findVersion(commitIds);
    }

    @Override
    public Version findVersion(List<String> commitIds) throws URISyntaxException, IOException, UnsuccessfulRequestException {
        VersionSearchModel searchModel = new VersionSearchModel(commitIds);
        UncommitedVersionModel version = httpHelper.findVersion(searchModel);
        return new Version(version.getVersion());
    }

    @Override
    public Version claimVersion(List<String> commitIds, String message, String headCommitId) throws IOException, UnsuccessfulRequestException {
        VersionCreateModel versionCreateModel = new VersionCreateModel(commitIds, message, headCommitId, configuration.getToken());
        VersionCommitModel commitedVersionModel = httpHelper.claimVersion(versionCreateModel);
        return new Version(commitedVersionModel.getVersion());
    }

    @Override
    public Version claimVersion(File rootProjectDir) throws IOException, GitAPIException, UnsuccessfulRequestException {
        GitHelper gitHelper = new GitHelper(rootProjectDir);
        List<String> commitIds = gitHelper.findCommitsFromHead();
        return claimVersion(commitIds, gitHelper.getHeadCommitMessage(), gitHelper.getHeadCommitId().getName());
    }

    @Override
    public RepoResponseModel createRepository(RepoCreateModel repoCreateModel) throws IOException, UnsuccessfulRequestException {
        return httpHelper.createRepo(repoCreateModel);
    }
}

File: clients/java-client/src/main/kotlin/io/ehdev/conrad/client/java/internal/DefaultVersionManagerClient.kt
package io.ehdev.conrad.client.java.internal

import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.module.kotlin.KotlinModule
import io.ehdev.conrad.client.java.*
import io.ehdev.conrad.model.commit.CommitIdCollection
import io.ehdev.conrad.model.repository.CreateRepoRequest
import io.ehdev.conrad.model.repository.CreateRepoResponse
import io.ehdev.conrad.model.version.CreateVersionRequest
import okhttp3.OkHttpClient
import org.eclipse.jgit.api.errors.GitAPIException
import retrofit2.Retrofit
import retrofit2.converter.jackson.JacksonConverterFactory
import java.io.File
import java.io.IOException
import java.net.URISyntaxException

class DefaultVersionManagerClient(val client: OkHttpClient, val configuration: VersionServiceConfiguration) : VersionManagerClient {

    private val conradClient: ConradClient

    init {
        val mapper = ObjectMapper().registerModule(KotlinModule())

        val retrofit: Retrofit = Retrofit.Builder()
            .baseUrl(configuration.applicationUrl)
            .addConverterFactory(JacksonConverterFactory.create(mapper))
            .client(client)
            .build();

        conradClient = ConradClient(retrofit.create(ProjectConradClient::class.java), retrofit.create(RepositoryConradClient::class.java))
    }

    @Throws(IOException::class, GitAPIException::class, URISyntaxException::class)
    override fun findFinalVersion(searchDir: File): Version {
        val gitHelper = GitHelper(searchDir)
        return Version(
            conradClient.findVersion(
                configuration.authToken!!,
                configuration.projectName!!,
                configuration.repoName!!,
                gitHelper.headCommitId.name).execute().body().version)
    }

    @Throws(IOException::class, GitAPIException::class, URISyntaxException::class)
    override fun findVersion(searchDir: File): Version {
        val gitHelper = GitHelper(searchDir)
        val commitIds = gitHelper.findCommitsFromHead()
        return findVersion(commitIds)
    }

    @Throws(URISyntaxException::class, IOException::class)
    override fun findVersion(commitIds: List<String>): Version {
        return Version(
            conradClient.findVersion(
                configuration.authToken!!,
                configuration.projectName!!,
                configuration.repoName!!,
                CommitIdCollection(commitIds)).execute().body().version)
    }

    @Throws(IOException::class)
    override fun claimVersion(commitIds: List<String>, message: String, headCommitId: String): Version {
        val newVersion = conradClient.createNewVersion(
            configuration.authToken!!,
            configuration.projectName!!,
            configuration.repoName!!,
            CreateVersionRequest(commitIds, message, headCommitId)).execute().body()

        return Version(newVersion.version)
    }

    @Throws(IOException::class, GitAPIException::class)
    override fun claimVersion(rootProjectDir: File): Version {
        val gitHelper = GitHelper(rootProjectDir)
        val commitIds = gitHelper.findCommitsFromHead()
        return claimVersion(commitIds, gitHelper.headCommitMessage, gitHelper.headCommitId.name)
    }

    @Throws(IOException::class)
    override fun createRepository(repoCreateModel: CreateRepoRequest): CreateRepoResponse {
        return conradClient.createRepository(
            configuration.authToken!!,
            configuration.projectName!!,
            configuration.repoName!!,
            repoCreateModel).execute().body()
    }

    @Throws(IOException::class, GitAPIException::class)
    override fun tagVersion(rootProjectDir: File, version: String) {
        GitHelper(rootProjectDir).tag(version)
    }
}
